Add Dyna. XTC sampler

CasualAutopsy opened the pull request:
A sampler adjustment I made back when XTC first released. Didn't make a PR at the time 'cus while it was a good improvement to make the sampler less schizo and more adherent, it wasn't up to my standards to be PR worthy. Now that Top N Sigma has been added however, it's more up to par when used along side it.

A good model to test it with is [MarinaraSpaghetti/NemoMix-Unleashed-12B](https://huggingface.co/MarinaraSpaghetti/NemoMix-Unleashed-12B) as it handles the N Sigma sampler ***very*** well.

Tried to add the sampler to klite myself this time, but I couldn't add it 100% perfectly (Red text appeared under the 3 complex samplers stating "X is not supported" for some reason) but I at least laid down some of the groundwork this time.

CasualAutopsy committed:
add dyna xtc and fix comment consistancy to appease ocd

CasualAutopsy committed:
expose dyna xtc in api

CasualAutopsy committed:
add dyna xtc to klite

LostRuins commented:
Hi and thanks for the PR. 

The original XTC is intended to increase creativity by removing common tokens, while maintaining coherence by having a threshold cutoff. 

What is this change to XTC aimed to achieve?  Could you explain the rationale behind these changes? On first glance you seem to be just removing a ratio of a few additional selected tokens arbitrarily. 

Also tagging @p-e-w if they're interested to take a look.

EquinoxPsychosis commented:
> Hi and thanks for the PR.
> 
> The original XTC is intended to increase creativity by removing common tokens, while maintaining coherence by having a threshold cutoff.
> 
> What is this change to XTC aimed to achieve? Could you explain the rationale behind these changes? On first glance you seem to be just removing a ratio of a few additional selected tokens arbitrarily.
> 
> Also tagging @p-e-w if they're interested to take a look.

The goal of this change is to find a middle ground of both creativity and adherence. While the original does give a massive boost in creativity, adherence takes a massive hit, ruining a lot of a models ability to follow prompts. This change is meant to fix that problem.

EquinoxPsychosis commented:
So as a way to fix the adherence problem I thought it would be a good Idea to use division to determine how many tokens get trimmed. with an `xtc_divide` of 2, half of the tokens minus 1 would get trimmed.

For the sake of not fucking up the loops I used rounding, using `static_cast<...>()` to turn the calculated value into `size_t` and during that, I had the thought 'Why not let the user decide how rounding should be done' and so I made it an option as well with `xtc_round` giving the user the choice of floor, ceil, or regular rounding.

EquinoxPsychosis commented on gpttype_adapter.cpp:
```diff
@@ -913,54 +913,82 @@ void sample_top_a(llama_token_data_array * candidates, float a, size_t min_keep)
+                candidates->data[i].logit -=999.0f;  // infinity gets wonky results downstream, this hack works well enough
+            }
+        }
+        else { // otherwise xtc sticks to default trimming behaviour
```
As you can see, for the sake of compatibility for older presets, I've kept the old trimming behaviour, and Dyna XTC is disabled by default. `xtc_divide` must be greater than 1 for it to use the new way of trimming.

LostRuins replied:
That seems somewhat counterproductive though, XTC already provides the tools requires to increase or reduced the trimming threshold i.e. `xtc_threshold`, wouldn't that solve the issue?

Also, is @EquinoxPsychosis and @CasualAutopsy  the same person?

CasualAutopsy replied:
> That seems somewhat counterproductive though, XTC already provides the tools requires to increase or reduced the trimming threshold i.e. `xtc_threshold`, wouldn't that solve the issue?

I wouldn't say so. For starters it allows much more smoother control over it than just simply adjusting threshold.

It can also allow to keep strong creativity during token generations when there is medium to large amounts of entropy, and when there's low entropy, it will stick to trimming less, keeping things adherent when it has to be, and creative when it needs to be.

For example if you set the threshold to 25, it will indeed solve the adherence issue, but if there is high entropy where the probabilities are spread to thinly across the tokens, it will ruin the creativity, so I created this to give a balance of both.

So when the AI is confident, it will remain confident, adhering to things predefined in the context, and when it isn't confident it will generate something more out of pocket.

CasualAutopsy replied:
> Also, is @EquinoxPsychosis and @CasualAutopsy the same person?

And yes, we are both the same person. This is so I can have a branch of kobold without deleting the branch of llama.cpp that I have

LostRuins commented:
Alright I guess I see what you are getting at.

I think we can simplify it though. What about a single extra parameter `xtc_ratio` which varies between 0.0 (no effect) to 1.0 (full strength) which affects how many tokens get pruned. Defaults to 1.0 aka current behavior of full XTC strength. That seems clearer than `xtc_divide`. So if XTC would have removed 10 tokens and `xtc_ratio` is 0.6, then it would remove 6 tokens instead.

I would also suggest getting rid of `xtc_round` for simplicity, simply pick the best one (closest?) and stick to that. 

P.s. dont worry about editing the frontend ui, i can patch that in later.

p-e-w commented:
Hmm. To be frank, this change seems rather ad-hoc.

The *number* of tokens removed means essentially nothing without taking the shape of the probability distribution into account, and limiting how many tokens are pruned by XTC doesn't affect the outcome in any predictable way. This is also why Top-K is such a poor truncation sampler: Performing operations based on token indices is highly context-dependent and doesn't make any tangible guarantees about the resulting distribution.

By contrast, XTC's control parameters are designed to provide quite specific guarantees. XTC ensures that at least one token with probability >= `xtc_threshold` remains in the output distribution. That's very stable and easy to understand. If you consider a probability of 20% to be "reasonable", you can simply set `xtc_threshold` to 0.2, and XTC will always leave at least one reasonable token. I fail to see how the proposed parameters translate to anything resembling an intuitive control mechanism. They may provide subjectively better responses in some cases, while rendering XTC ineffective in others, but which of the two happens is essentially a coin toss.

XTC already has two parameters for controlling its strength. As `xtc_threshold` approaches 0.5, the effect of XTC vanishes, and as `xtc_probability` approaches 0, the effect of XTC also vanishes. Control is completely continuous along two orthogonal dimensions. This means that for a given effect strength, you actually have a 1-dimensional control manifold of constant effect strength that allows you to additionally vary the effect "flavor". That's quite a lot of control IMO.

EquinoxPsychosis commented:
Given what p-e-w said I think I want to hold off on merging this. Not that I given up on the Dyna XTC idea *just* yet, but I want to give it a few tries at finding a way to better calculate things dynamically so that isn't *ONLY* good for the N Sigma sampler chain. I'll do it in a separate branch so if I'm able to figure something out, I'll add it to this PR, if not then I think I'll close it.

LostRuins commented:
Alright then, no worries.

CasualAutopsy committed:
Revert "add dyna xtc to klite"

This reverts commit 90e96ed5c2a5d7d72625288831313b9f86ec7532.

CasualAutopsy committed:
dyna xtc nsigma test

CasualAutopsy committed:
add nsigma based dyna xtc

CasualAutopsy committed:
fix

CasualAutopsy commented:
how... how does merging two simple truncation samplers give such amazing results...?
I've no words other than simple instructions:

When using it with nsigma, make sure you set xtc's nsigma to a lower value.
Recommended starting values is `1.0` for `Nsigma` and `0.5` for `XTC Nsigma`

if using NemoMix Unleased, I suggest the starting values of `60` for `Top K` and `Dyna. Temp.` with the following:
`1` for Min
`4` for Max
`0.85` for Exp

when using it on its own then make sure to use at least one truncation sampler, otherwise outputs are incoherent. `Min P` of  `0.1` and `XTC Nsigma` of `0.25` is a good starting point

CasualAutopsy committed:
newline cleanup

CasualAutopsy committed:
split trimming methods

CasualAutopsy committed:
newline cleanup

CasualAutopsy committed:
screw off VS

CasualAutopsy committed:
VS, why are you like this

CasualAutopsy committed:
split smooth sampling for sake of nsigma compatability
- smooth sampling uses soft max, so it shouldn't be used before nsigma

CasualAutopsy committed:
add min exclude token limit

EquinoxPsychosis commented on gpttype_adapter.cpp:
```diff
@@ -1647,15 +1687,66 @@ const std::vector<samplers> & sampler_order, llama_grammar * grammar, float dyna
+        } else {
+            sample_temperature(&candidates_p, temp);
+        }
+        xtc_nsig_tokens = exclude_n_sigma(&candidates_p, xtc_nsigma);
```
It's cool that my idea to do this actually worked. Makes me wounder if it's possible to do with other samplers to break past their fixed order, like DRY for example.

Don't want to bother Lost with too many sampler PRs past Dyna. XTC, so I'm going to make my own fork like Nexesenex did to test and play around with my sampler ideas. I'm ceratainly going to try and see If I can use this method to break DRY's locked ordering at some point

CasualAutopsy committed:
add comments and make some pre existing ones more consistent

CasualAutopsy committed:
add comments and make some pre existing ones more consistent

CasualAutopsy committed:
Merge branch 'com_exp/Dyna_XTC' of https://github.com/EquinoxPsychosis/koboldcpp into com_exp/Dyna_XTC

CasualAutopsy commented:
~~might seem a bit silly to set it to draft for just comments, but I was seeing if I could do any refactoring, and remove the freq_pen calculations from the xtc nsigma mask. both were above my skill level so I just added some comments for readability and better understanding~~

Edit: My scatter brained ass forgot we're in the middle of the AI boom, so I could've gotten an LLM to help with it. Putting the PR back in draft

CasualAutopsy committed:
Refactoring
- refactoring for `exclude_n_sigma`
- make `exclude _n_sigma` follow more in lines with PR #1394 (For real this time)

CasualAutopsy committed:
Refactor pt2: `sample_xtc`
- refactor dynaxtc changes to `sample_xtc`
- add/revise some comments

CasualAutopsy committed:
Merge branch 'concedo_experimental' into com_exp/Dyna_XTC

CasualAutopsy commented:
Closing PR since I've made my own KCpp fork to create, test, and experiment with custom samplers.
If you're curious about this sampler you can try it out at the EquinoxPsychosis/sampboldcpp repo.
